App switcher

jdlrobson opened the pull request:
This adds some code to TiddlySpace for the app switcher.

Note this pull request does not do the following:
1) Surface the app switcher in TiddlyWiki or any other app.
2) Implement what goes in the app switcher. (http://apps.tiddlyspace.com/app_picker has been suggested but has not been included in this pull)

It also assumes the TiddlySpace icon, however this may change at a future date

To test add 

```
<script type="text/javascript" src="/bags/apps/tiddlers/ts-application"></script>
```

to your MarkupPostHead tiddler in the TiddlyWiki serialization.

jdlrobson committed:
add apps bag with app switcher code

The idea is an app on tiddlyspace can include
<script type="text/javascript" src="/bags/apps/tiddlers/ts-application.js"></script>

and it will provide a way of switching between apps on TiddlySpace

it loads a file 'app-picker.html' into an iframe which is yet to be defined

it is important this content lives in a separate bag that is not included
in any recipe so that it doesn't clutter up existing users TiddlyWiki's

jdlrobson committed:
make apps back available via control view

its important this bag is accessible from any domain on tiddlyspace
so that the app switcher can be loaded into any app reliably

jdlrobson committed:
add bubble argument to addEventListener calls

Some versions of Firefox require a 3rd argument for
addEventListener otherwise will throw exception.

jdlrobson committed:
stop overlap between icon and bubble

this is how speech bubbles tend to work - with a gap between the bubble
and the 'speaker'

cdent commented:
Some questions for understanding:
- I assume the "apps" bag is for default apps that we intend to include in the distribution?
- What does "ts-application" mean as a name? I've had several interpretations cross my mind, none of which seem to work: "tiddlyspace application", "task switcher application", or perhaps "tiddlyspace application <chooser>". In 1 and 3 tiddlyspace is redundant. In 2 "application" is redundant. What was wrong with "app switcher"?

In general we should avoid the ts- prefix, I think. We should also avoid [Hungarian Notation](http://en.wikipedia.org/wiki/Hungarian_notation).

jdlrobson commented:
I agree, the names all need fixing. I'm not fussed either way. AppSwitcher could work, likewise backstage or launchpad or launcher.

The apps bag could be used for default apps included in the distribution, however I suspect we would use inclusion instead as a method of packaging them. Remember currently tiddlers in the apps bag wouldn't be accessible via friendly uris
e.g. /takenote would become /bags/apps/tiddlers/takenote which is not great.

I thought about the name lib for the apps bag but then this is not true either, as css and html is required.
The bag name might possibly be the name we give to the product.
e.g.
/bags/applauncher

So I guess we should hold off pulling this until these names are finalised.

jdlrobson committed:
change bubble color

feedback from colm to change colour of bubble

cdent commented:
So, if the apps bag is not for apps, what is it for?

If the plan is to use inclusion to present apps, then there needs to be an apps space created and the necessary infrastructure for managing its inclusion.

Overall I'm confused about the several purposes of this pull? Even if the names are finalized it still seems like the apps bag is a question mark.

jdlrobson commented:
1) The code needs to be available across all spaces without control view and shouldn't be excludable.
2) The code shouldn't be included into existing spaces (as the expectation is apps will have a script tag point to the javascript file)
3) This code is NOT an app. It is a bit of javascript that provides a way to navigate to apps in a given space.

So based on this we could use the bag common, however it seems anyone can write to the common bag? It also feels like it should live in a bag with a more meaningful name.

Does this make sense?

I suggest not pulling yet, as things will make more sense hopefully with further commits tomorrow...

jdlrobson commented:
Talking with Jeremy now, it has been suggested that it might be packaged in the tiddlyspace bag as this allows users to explore the data (although I'm not sure this is a good thing). It also removes the need for a script tag with a simple systemConfig tag.

Personally I think we should move away from this and I worry that if it is baked into the TiddlyWiki as a plugin, we are allowing users to overwrite the tiddler and break the consistency across spaces we are striving to get to.

cdent commented:
The diversity of opinions on these matter makes it sound like a pull request is not the place to be having the discussion.

Most clearly:
- The app switcher should not be a tiddlywiki plugin because clearly we want to be able to load the app switcher from apps.

My expectation is that apps themselves are HTML files that point to one or more javascript and css files. That's not germane to this discussion, really.

Using common might work, common is not supposed to be writable by everyone, that's a bug (now #724). If the apps bag is just for this small set of tiddlers, then common is better. If there will be more, then okay.

I agree that this stuff should be explorable, but the truth is that anything that is readable is explorable, because of the web api and we _do not_ want to add more than we have to to the list of tiddlers generated when processing a person's recipe, so tiddlyspace is not great.

So common seems good, or perhaps "appsupport" if there is going to be lots of such things. 

Something to keep in mind is that a bag with lots of tiddlers in it is not bad if those tiddlers are only being accessed by name (rather than the whole collection). Since these things we're talking about are things which will be accessed by tiddler is okay.

jdlrobson committed:
use common bag instead of apps

on second thought and after review with @cdent it seems that the
common bag should be sufficient for these tiddlers.

It shouldn't be included in the space and should be within the control
view scope

jdlrobson committed:
rename to backstage

the backstage name is more meaningful and sets an expectation that
this will eventually replace the tiddlywiki backstage as the place
to go to manage a space

note backstage.js.js as the name of this tiddler should be "backstage.js"
not "backstage"

jdlrobson committed:
update links in backstage.js

make them point to common bag tiddlers

jdlrobson committed:
package up jquery.js and ts.js in the common bag

many apps are now starting to use them, and the app page / backstage bubble
will use both of them.

jdlrobson committed:
add backstage html

this defines the page "backstage" that loads in this frame
it also provides the css required to style it and some images it
makes use of

jdlrobson committed:
box shadow for other browsers

use other vendor prefixes

jdlrobson committed:
increase bubble height

jdlrobson committed:
correct tail colors of bubble

they are wrong

jdlrobson committed:
reduce size of arrows

they are a bit too big and overwhelming

jdlrobson committed:
form styling for backstage

jdlrobson committed:
add anchor styling

when logged in message "welcome back jon" appears
jon links to space for jon
no styling at moment.
until now.

jdlrobson commented:
This has now been updated and I believe is ready to go!
